temporal/api/cloud/namespace/v1/message.proto

```diff
@@ -0,0 +1,120 @@
+    // Optional, default is empty.
+    repeated string passive_regions = 1;
+}
+
+message NamespaceSpec {
+    // The name to use for the namespace.
+    // This will create a namespace that's available at '<name>.<account>.tmprl.cloud:7233'.
+    // The name is immutable. Once set, it cannot be changed.
+    string name = 1;
+    // The region where the namespace is (or will be) active.
+    // The region is immutable. Once set, it cannot be changed.
+    string region = 2;
+    // The number of days the workflows data will be retained for.
+    // Changes to the retention period may impact your storage costs.
+    // Any changes to the retention period will be applied to all new running workflows.
+    int32 retention_days = 3;
+    // The mtls authentication and authorization to enforce on the namespace.
+    MtlsAuthSpec mtls_auth = 4;
+    // The custom search attributes to use for the namespace.
+    // The name of the attribute is the key and the type is the value.
+    // Supported attribute types: text, keyword, int, double, bool, datetime, keyword_list.
+    // NOTE: currently deleting a search attribute is not supported.
+    // Optional, default is empty.
+    map<string, string> custom_search_attributes = 5;
+    // Codec server spec used by UI to decode payloads for all users interacting with this namespace.
+    // Optional, default is unset.
+    CodecServerSpec codec_server = 6;
+    // The global namespace spec.
+    // Optional, default is unset.
+    GlobalSpec global = 7;
+}
+
+message Endpoints {
+    // The web ui address.
+    string web_address = 1;
+    // The grpc hostport that the temporal workers, clients and tctl connect to.
+    string grpc_hostport = 2;
+}
+
+message OperatingEnvelope {
+    // The namespace may be throttled if its APS exceeds the limit.
+    int32 actions_per_second_limit = 1;
+}
+
+message AWSPrivateLink {
+    // The list of principals that are allowed to access the private links on the namespace.
+    repeated string allowed_principal_arns = 1;
+    // The list of private links, if any, that are associated with the namespace.
+    repeated string vpc_endpoint_service_names = 2;
+}
+
+message Namespace {
+    // The namespace identifier.
+    string namespace = 1;
+    // The current version of the namespace specification.
+    // The next update operation will have to include this version.
+    string resource_version = 2;
+    // The namespace specification.
+    NamespaceSpec spec = 3;
+    // The current state of the namespace.
+    string state = 4;
+    // The id of the async operation that is creating/updating/deleting the namespace, if any.
+    string async_operation_id = 5;
+    // The endpoints for the namespace.
+    Endpoints endpoints = 6;
+    // The envelope is a list of service level agreements (SLAs) that can be provided around a given namespace.
+    OperatingEnvelope operating_envelope = 7;
+    // The private link for the namespace, if any.
+    oneof private_link {
```

Comment:
In the future, its possible that we have a global namespace across clouds which uses both AWS private link and GCP private service connect (as an example). Or wireguard etc.

We may want to drop the `oneof` to ensure we aren't limited in the protocol.

Reply:
While it is a protobuf backwards compatible operation to move a field into a `oneof` later, it's often not a code-gen backwards compatible alteration. I do think it should be decided what compatibility guarantees you want to provide when you do provide them - wire guarantees or code-gen guarantees. If only the former, which is reasonable, then yes can wait on the `oneof`.